File: cpp/main/src/parser.cpp
/**
 * @file parser.cpp
 * @brief Implementation of the Parser class for Dragonfly Markup Language (DFML) parsing.
 * @author Javier Candales (javier_candales@yahoo.com.ar)
 * @date 2024-01-20
 * 
 * @copyright Copyright (c) 2024
 */

#include <dfml/parser.h>
#include <dfml/node.h>
#include <dfml/data.h>
#include <dfml/comment.h>

namespace dfml {

/**
 * @brief Constructor for the Parser class.
 * @param data The DFML data to be parsed.
 */
Parser::Parser(const std::string data) {
	i.set_data(data);
}

/**
 * @brief Static factory method to create an instance of the Parser class.
 * @param data The DFML data to be parsed.
 * @return A shared pointer to the created Parser object.
 */
std::shared_ptr<Parser> Parser::create(const std::string data) {
	return std::make_shared<Parser>(data);
}

/**
 * @brief Parses the DFML data and returns a list of shared pointers to parsed elements.
 * @return A list of shared pointers to parsed elements.
 */
std::list<std::shared_ptr<Element>> Parser::parse() {
	std::list<std::shared_ptr<Element>> list;
	
	parse_children(list);

	return list;
}

/**
 * @brief Parses child elements in the DFML data.
 * @param childs Reference to a list to store the parsed child elements.
 */
void Parser::parse_children(std::list<std::shared_ptr<Element>> &childs) {
	int ch;
	dfml::Value value;
	while ((ch = i.next()) != -1) {
		switch (ch) {
		case ' ':
		case '\t':
		case '\n':
			break; // space (continue)

		case '/':
		case '#':
			i.back();
			childs.push_back(parse_comment());
			break;

		case '"':
		case '\'':
			parse_string(value);
			childs.push_back(dfml::Data::create(value));
			break;
		
		// End of parsing chidren
		case '}': return ;
		
		default:
			if (std::isalpha(ch)) {
				i.back();
				childs.push_back(parse_node());
			} else if (std::isdigit(ch)) {
				i.back();
				parse_number(value);
				childs.push_back(dfml::Data::create(value));
			} else {
				throw ParserException("Invalid character for node child on line: " +
						i.get_line());
			}
		}
	}
}

/**
 * @brief Parses a node element in the DFML data.
 * @return A shared pointer to the parsed node element.
 */
std::shared_ptr<Element> Parser::parse_node() {
	int ch;
	std::string name = parse_node_name();

	// If keywords "true" or "false" isn't a node: it is boolean data.
	if (name == "true") {
		return dfml::Data::create_boolean(true);
	} else if (name == "false") {
		return dfml::Data::create_boolean(false);
	}

	// Create a node
	auto node = dfml::Node::create(name);
	std::list<std::shared_ptr<Element>> children;

	if (i.end()) return node;

	i.back();

	if (node->get_name() == "") {
		throw ParserException("Empty node name encountered on line: " + i.get_line());
	}

	// Parse attributes and children
	bool stop = false, attr_parsed = false;
	while ((ch = i.next()) != -1) {
		switch (ch) {
		case ' ':
		case '\t':
		case '\n':
			break; // space (continue)

		case '(':
			if (attr_parsed) {
				throw ParserException("Double attribute list found in the node on line: " +
						i.get_line());
			}
			parse_node_attributes(node);
			attr_parsed = true;
			break;

		case '{':
			parse_children(children);
			stop = true;
			break;

		case '}':
			stop = true;
			break;
		default:
			stop = true;
			i.back();
			break;
		}

		if (stop) break;
	}

	for (auto &e : children) {
		node->add_child(e);
	}

	return node;
}

/**
 * @brief Parses the name of a node element in the DFML data.
 * @return The name of the parsed node.
 */
const std::string Parser::parse_node_name() {
	int ch;
	std::string name = "";
	while ((ch = i.next()) != -1) {
		if (std::isalnum(ch)) {
			name += ch;
		} else break;
	}
	return name;
}

/**
 * @brief Parse attibutes for the given node.
 * @param node The node reference.
 * 
 */
void Parser::parse_node_attributes(std::shared_ptr<Node> node) {
	int ch;
	bool stop = false;
	
	while ((ch = i.next()) != -1 && !stop) {
		switch (ch) {
		case ' ':
		case '\n':
		case '\t':
		case '\r':

		case ',': // TODO: Improve ',' separator
			// Continue
			break;

		case ')':
			stop = true;
			break;
		}

		if (std::isalpha(ch)) {
			i.back();
			parse_node_attribute(node);
		}

		if (stop) break;
	}
}

/**
 * @brief Parse a attribute pair (Key/Value) for the given node.
 * @param node The node reference
 */
void Parser::parse_node_attribute(std::shared_ptr<Node> node) {
	int ch;
	bool stop = false;
	std::string key;
	Value value;

	enum status_t {
		PARSING_NAME,
		FIND_SEP,
		FIND_VALUE
	} status = PARSING_NAME;

	// Parse key
	while ((ch = i.next()) != -1 && !stop) {
		switch (status) {
		case FIND_VALUE:

			switch (ch) {

			case ' ':
			case '\n':
			case '\t':
			case '\r':
				// Continue
				break;

			case '"':
			case '\'':
				parse_string(value);
				node->set_attribute(key, value);
				return ;
			
			/*default:
				// Error
				break;*/
			}

			if (is_number(ch)) {
				i.back();
				parse_number(value);
				node->set_attribute(key, value);
				i.back();
				return;
			}
			if (std::isalpha(ch)) {
				i.back();
				parse_boolean(value);
				node->set_attribute(key, value);
				i.back();
				return;
			}

			break;
		case FIND_SEP:
			switch (ch) {
			case ' ':
			case '\n':
			case '\t':
			case '\r':
				// Continue
				break;

			case ':':
				status = FIND_VALUE;
				break;
			case ',':
			case ')':
				// Empty attribute
				node->set_attr_string(key, "");
				return;
			}
			break;

		case PARSING_NAME:
			switch (ch) {

			case ' ':
			case '\n':
			case '\t':
			case '\r':
			case ':':
				status = FIND_SEP;
				i.back();
				break;

			case ',':
			case ')':
				// Empty attribute
				node->set_attr_string(key, "");
				return;

			/*default:
				// Error
				return;*/
			}


			if (std::isalnum(ch) || ch == '_') {
				key += ch;
			}
			break;
		}
	}
}

/**
 * @brief Parses a string element in the DFML data.
 * @param value Value reference to set string data.
 */
void Parser::parse_string(dfml::Value &value) {
	std::string result;
	int ch;

	int end = i.current();

	while ((ch = i.next()) != -1) {
		if (ch == end) break;
		result += ch;
	}
	
	value.set_string(result);
}

/**
 * @brief Parses a number Data element.
 * Detect float point to se double data or se fixed to integer.
 * @param value Value reference to set number data.
 */
void Parser::parse_number(dfml::Value &value) {
	int ch;
	std::string result;
	double dbl_result = 0.0;
	long int_result = 0;
	bool dbl = false;
	size_t pos;
	bool error = false;

	while ((ch = i.next()) != -1) {
		if (!is_number(ch)) break;
		if (ch == '.') dbl = true;
		result += ch;
	}

	if (dbl) {
		try {
			dbl_result = std::stod(result, &pos);
			if (pos != result.length()) {
				error = true;
			}
		} catch( ... ) {
			error = true;
		}

		if (error) throw ParserException("Double conversion error on line: " + i.get_line());

		value.set_double(dbl_result);
	} else {
		try {
			int_result = std::stol(result, &pos);
			if (pos != result.length()) {
				error = true;
			}
		} catch( ... ) {
			error = true;
		}

		if (error) throw ParserException("Integer conversion error on line: " + i.get_line());

		value.set_integer(int_result);
	}
}

/**
 * @brief Parses a boolean Data element.
 * @param value Value reference to set boolean data.
 */
void Parser::parse_boolean(Value &value) {
	int ch;
	std::string result;
	while ((ch = i.next()) != -1) {
		if (!std::isalpha(ch)) break;
		result += ch;
	}

	if (result == "true" || result == "false") {
		value.set_boolean(result == "true" ? true : false);
	} else {
		throw ParserException("Boolean conversion error on line: " + i.get_line());
	}
}

/**
 * @brief Parses a Comment element.
 * Parses //, /* and # comment type.
 * 
 * @return std::shared_ptr<Element> The parsed Comment element.
 */
std::shared_ptr<Element> Parser::parse_comment() {
	int ch = i.next();
	bool single_line = false;
	std::string string;

	if (ch == '#') single_line = true;
	else if (ch == '/') {
		ch = i.next();
		if (ch == -1) {
			throw ParserException("Unexpected comment termination on line: " + i.get_line());
		} else if (ch == '/') {
			single_line = true;
		}
		else if (ch == '*') {single_line = false;}
		else {
			throw ParserException("Unexpected comment termination on line: " + i.get_line());
		}
	}

	bool stop = false;
	while ((ch = i.next()) != -1 && !stop) {
		switch(ch) {

		case '\r':
			if (!single_line) string += ch;
			// Continue
			break;

		case '\n':
			if (single_line) {
				i.back();
				stop = true;
			}
			else string += ch;
			break;

		case '*':
			if (!single_line) {
				ch = i.next();
				if (ch == '/' || ch == -1) {
					stop = true;
				} else {
					string += ch;
				}
			} else string += ch;
			break;

		default:
			string += ch;
		}
	}

	return dfml::Comment::create(string);
}

/**
 * @brief Checks if the given character represents a number.
 * @param ch The character to be checked.
 * @return True if the character represents a number, otherwise false.
 */
const bool Parser::is_number(int ch) const {
	if (std::isdigit(ch)) return true;
	if (ch == '-' || ch == '.') return true;
	return false;
}

/**
 * @brief Gets the next character in the DFML data.
 * @return The ASCII value of the next character, or -1 if end of data is reached.
 */
int CharIterator::next() {
	if (i >= data.size()) return -1;
	int ch = data[i]; i ++;
	if (ch == '\n') line ++;
	return ch;
}

/**
 * @brief Moves the iterator back by one character.
 */
void CharIterator::back() {
	i --;
}

}  // namespace dfml

File: js/main/parser.js
import { DFMLNode } from "./node.js"
import { DFMLData } from "./data.js"
import { DFMLValue } from "./value.js";
import { DFMLComment } from "./comment.js"
import { DFMLElement } from "./element.js";

/**
 * Exception class for parser-related errors.
 *
 * @class DFMLParserException
 * @extends {Error}
 */
class DFMLParserException extends Error {
	/**
	 * Creates an instance of DFMLParserException.
	 * @param {string} message message of the exception.
	 * @memberof DFMLParserException
	 */
	constructor(message) {
		super(message);
		this.name = "DFMLParserException";
	}
}

/**
 * Iterator for characters used by the Parser to iterate over a string.
 *
 * @class CharIterator
 */
class CharIterator {
	/**
	 * Creates an instance of CharIterator.
	 * @param {string} [data=''] string data.
	 * @memberof CharIterator
	 */
	constructor(data = '') {
		this.data = data;
		this.i = 0;
		this.line = 1;
	}

	/**
	 * Sets the data for iteration.
	 *
	 * @param {string} data The string data to iterate over.
	 * @memberof CharIterator
	 */
	setData(data) {
		this.data = data;
		this.i = 0;
		this.line = 1;
	}

	/**
	 * Retrieves the next character in the iteration.
	 *
	 * @return {number} The ASCII value of the next character.
	 * @memberof CharIterator
	 */
	next() {
		if (this.i >= this.data.length) return -1;
		const ch = this.data[this.i++];
		if (ch === '\n') this.line++;
		return ch.charCodeAt(0);
	}

	/**
	 * Retrieves the current character in the iteration.
	 *
	 * @return {number} The ASCII value of the current character.
	 * @memberof CharIterator
	 */
	current() {
		return this.data[this.i - 1];
	}

	/**
	 * Moves the iterator back to the previous character.
	 *
	 * @memberof CharIterator
	 */
	back() {
		this.i--;
	}

	/**
	 * Checks if the end of the iteration is reached.
	 *
	 * @return {boolean} True if the end is reached, false otherwise.
	 * @memberof CharIterator
	 */
	end() {
		return this.i >= this.data.length;
	}

	/**
	 * Returns current data line.
	 *
	 * @return {string} the current data line.
	 * @memberof CharIterator
	 */
	getLine() {
		return this.line.toString();
	}
}

/**
 * Class responsible for parsing DFML data and creating DFML objects.
 *
 * @export
 * @class DFMLParser
 */
export class DFMLParser {
	/**
	 * Creates an instance of DFMLParser.
	 * @param {string} data The DFML data to parse.
	 * @memberof DFMLParser
	 */
	constructor(data) {
		this.i = new CharIterator(data);
	}

	/**
	 * Creates and returns a DFMLParser instance.
	 *
	 * @static
	 * @param {string} data The DFML data to parse.
	 * @return {DFMLParser} the new DFMLParser instance.
	 * @memberof DFMLParser
	 */
	static create(data) {
		return new DFMLParser(data);
	}

	/**
	 * Parses the DFML data and returns a list of parsed DFMLElement objects.
	 *
	 * @return {Array.<DFMLElement>} The list of parsed DFMLElement objects.
	 * @memberof DFMLParser
	 */
	parse() {
		const list = [];
		this.parseChildren(list);
		return list;
	}

	/**
	 *  Parses the children of a DFMLNode.
	 *
	 * @param {Array.<DFMLElement>} childs The list to store the parsed child elements.
	 * @memberof DFMLParser
	 */
	parseChildren(childs) {
		let ch;
		let value = new DFMLValue();
		while ((ch = this.i.next()) !== -1) {
			switch (ch) {
				case 32: // space
				case 9:  // tab
				case 10: // newline
				case 13: // carriage return 
					break;

				case 47: // '/'
				case 35: // '#'
					this.i.back();
					childs.push(this.parseComment());
					break;

				case 34: // '"'
				case 39: // "'"
					value = new DFMLValue();
					this.parseString(value);
					childs.push(DFMLData.createWithValue(value));
					break;

				case 125: // '}'
					return;

				default:
					//if (/[a-zA-Z]/.test(String.fromCharCode(ch))) {
					if (this.isAlpha(ch)) {
						this.i.back();
						childs.push(this.parseNode());
					} else if (/[0-9]/.test(String.fromCharCode(ch))) {
						this.i.back();
						value = new DFMLValue();
						this.parseNumber(value);
						childs.push(DFMLData.createWithValue(value));
					} else {
						throw new DFMLParserException("Invalid character for node child on line: " + this.i.getLine());
					}
			}
		}
	}

	/**
	 * Parses a DFMLNode element.
	 *
	 * @return {DFMLNode} The parsed DFMLNode element.
	 * @memberof DFMLParser
	 */
	parseNode() {
		let ch;
		const name = this.parseNodeName();

		if (name === "true") return DFMLData.createBoolean(true);
		if (name === "false") return DFMLData.createBoolean(false);

		const node = DFMLNode.create(name);
		let children = [];

		if (this.i.end()) return node;

		this.i.back();

		if (!node.name) {
			throw new DFMLParserException("Empty node name encountered on line: " + this.i.getLine());
		}

		let stop = false;
		let attrParsed = false;

		while ((ch = this.i.next()) !== -1) {
			switch (ch) {
				case 32:
				case 9:
				case 10:
				case 13:
					break; // space (continue)

				case 40: // '('
					if (attrParsed) {
						throw new DFMLParserException("Double attribute list found in the node on line: " + this.i.getLine());
					}
					this.parseNodeAttributes(node);
					attrParsed = true;
					break;

				case 123: // '{'
					this.parseChildren(children);
					stop = true;
					break;

				case 125: // '}'
					stop = true;
					this.i.back();
					break;

				default:
					stop = true;
					this.i.back();
					break;
			}

			if (stop) break;
		}

		children.forEach((c) => {
			node.addChild(c);
		});

		return node;
	}

	/**
	 * Parses the name of a DFMLNode element.
	 *
	 * @return {string} The parsed name of the DFMLNode element.
	 * @memberof DFMLParser
	 */
	parseNodeName() {
		let name = '';
		let ch;
		while ((ch = this.i.next()) !== -1) {
			//if (/[a-zA-Z0-9]/.test(String.fromCharCode(ch))) {
			if (this.isAlphaNum(ch)) {
				name += String.fromCharCode(ch);
			} else break;
		}
		return name;
	}

	/**
	 * Parse attibutes for the given node.
	 *
	 * @param {DFMLNode} node The node reference.
	 * @memberof DFMLParser
	 */
	parseNodeAttributes(node) {
		let ch;
		let stop = false;

		while ((ch = this.i.next()) !== -1 && !stop) {
			switch (ch) {
				case 32:
				case 9:
				case 10:
				case 13:
					break;

				case 41: // ')'
					stop = true;
					break;
			}

			//if (/[a-zA-Z]/.test(String.fromCharCode(ch))) {
			if (this.isAlpha(ch)) {
				this.i.back();
				this.parseNodeAttribute(node);
			}

			if (stop) break;
		}
	}

	/**
	 * Parse a attribute pair (Key/Value) for the given node.
	 *
	 * @param {DFMLNode} node The node reference.
	 * @memberof DFMLParser
	 */
	parseNodeAttribute(node) {
		let key = '';
		let value = new DFMLValue();
		let status = 'PARSING_NAME';
		let ch;

		while ((ch = this.i.next()) !== -1) {
			switch (status) {
				case 'FIND_VALUE':
					if (ch === 34 || ch === 39) { // '  "
						this.parseString(value);
						//node[key] = value;
						node.setAttribute(key, value);
					//} else if (/[0-9]/.test(String.fromCharCode(ch))) {
					} else if (this.isNumber(ch)) {
						this.i.back();
						this.parseNumber(value);
						node.setAttribute(key, value);
						this.i.back();
					//} else if (/[a-zA-Z]/.test(String.fromCharCode(ch))) {
					} else if (this.isAlpha(ch)) {
						this.i.back();
						this.parseBoolean(value);
						node.setAttribute(key, value);
						this.i.back();
					} else if (ch === 32 || ch === 10 || ch === 9 || ch === 8 || ch === 13) {

					} else if (ch === 44) { // ,
						return ;
					} else if (ch === 41) { // )
						this.i.back();
						return ;
					}

				case 'FIND_SEP':
					if (ch === 58) status = 'FIND_VALUE'; // ':'
					break;

				case 'PARSING_NAME':
					if (ch === 32 || ch === 10 || ch === 9 || ch === 8 || ch === 13
							|| ch === 58) { // :
						status = 'FIND_SEP';
						this.i.back();
					} else if (ch === 44 || ch === 41) { // ,  (
						// Empty attribute
						node.setAttrString(key, "");
						this.i.back();
						return ;
					//} else if (/[a-zA-Z0-9_]/.test(String.fromCharCode(ch))) {
					} else if (this.isAlphaNum(ch)) {
						key += String.fromCharCode(ch);
					}
					break;
			}
		}
	}

	/**
	 * Parses a string DFMLData element.
	 *
	 * @param {DFMLValue} value value to set string data.
	 * @memberof DFMLParser
	 */
	parseString(value) {
		let result = '';
		const end = this.i.current();
		let ch;
		while ((ch = this.i.next()) !== -1) {
			if (ch === end.charCodeAt(0)) break;
			result += String.fromCharCode(ch);
		}
		value.setString(result)
	}

	/**
	 * Parses a number DFMLData element.
	 * Autodetect float point to se double data or se fixed to integer.
	 * 
	 * @param {DFMLValue} value value to set number data.
	 * @memberof DFMLParser
	 */
	parseNumber(value) {
		let result = '';
		let ch;
		let dbl = false;

		while ((ch = this.i.next()) !== -1) {
			//if (!/[0-9.]/.test(String.fromCharCode(ch))) break;
			if (!this.isNumber(ch)) break;
			if (ch === 46) dbl = true;
			result += String.fromCharCode(ch);
		}

		if (dbl) {
			value.setDouble(parseFloat(result))
		} else {
			value.setInteger(parseInt(result, 10))
		}
	}

	/**
	 * Parses a boolean DFMLData element.
	 *
	 * @param {DFMLValue} value value to set boolean data.
	 * @memberof DFMLParser
	 */
	parseBoolean(value) {
		let result = '';
		let ch;
		while ((ch = this.i.next()) !== -1) {
			//if (!/[a-zA-Z]/.test(String.fromCharCode(ch))) break;
			if (!this.isAlpha(ch)) break;
			result += String.fromCharCode(ch);
		}
		if (result === "true" || result === "false") {
			value.setBoolean(result === "true");
		} else {
			throw new DFMLParserException("Boolean conversion error on line: " + this.i.getLine());
		}
	}

	/**
	 * Parses a DFMLComment element.
	 * Parses '//', '/*' and '#' comment type.
	 *
	 * @return {DFMLComment} The parsed DFMLComment element.
	 * @memberof DFMLParser
	 */
	parseComment() {
		let ch = this.i.next();
		let singleLine = false;
		let string = '';

		if (ch === 35) // #
			singleLine = true;
		else if (ch === 47) { // /
			ch = this.i.next();
			if (ch === -1)
				throw new DFMLParserException("Unexpected comment termination on line: " + this.i.getLine());
			else if (ch === 47) // /
				singleLine = true;
			else if (ch === 42) // *
				singleLine = false;
			else {
				throw new DFMLParserException("Unexpected comment termination on line: " + this.i.getLine());
			}
		}

		let stop = false;
		while ((ch = this.i.next()) !== -1 && !stop) {
			switch(ch) {
			case 13: // \r
				if (!singleLine) string += String.fromCharCode(ch);
				break;
			case 10: // \n
				if (singleLine) {
					this.i.back();
					stop = true;
				} else string += String.fromCharCode(ch);
				break;

			case 42: // *
				if (!singleLine) {
					ch = this.i.next();
					if (ch === 47 || ch === -1) // /
						stop = true;
					else
						string += String.fromCharCode(ch);
				} else string += String.fromCharCode(ch);
				break;

			default:
				string += String.fromCharCode(ch);
			}
		}

		return DFMLComment.createWithContent(string);
	}

	/**
	 * Checks the character ch if alphabetic, '-', or '_'.
	 *
	 * @param {string} ch character to check.
	 * @return {boolean} true if ch is 'A'-'Z', 'a'-'z', '-' or '_'.
	 * @memberof DFMLParser
	 */
	isAlpha(ch) {
		if (ch === 45) return true; // -
		if (ch === 95) return true; // _
		if (/[a-zA-Z]/.test(String.fromCharCode(ch))) return true;
		return false;
	}

	/**
	 * Checks the character ch if alphanumeric, '-', or '_'.
	 *
	 * @param {string} ch character to check.
	 * @return {boolean} true if ch is 'A'-'Z', 'a'-'z', '0'-'9', '-' or '_'.
	 * @memberof DFMLParser
	 */
	isAlphaNum(ch) {
		if (/[0-9]/.test(String.fromCharCode(ch))) return true;
		return this.isAlpha(ch);
	}

	/**
	 * Checks if the provided character represents a number.
	 *
	 * @param {string} ch The ASCII value of the character to check.
	 * @return {boolean} True if the character represents a number, false otherwise.
	 * @memberof DFMLParser
	 */
	isNumber(ch) {
		if (ch === 45) return true;
		if (ch === 46) return true;
		if (/[0-9]/.test(String.fromCharCode(ch))) return true;

		return false;
	}
}
